spark/src/test/java/org/apache/zeppelin/spark/SparkInterpreterTest.java

```diff
@@ -92,14 +96,40 @@ public void setUp() throws Exception {
       repl.open();
     }
 
+    final RemoteEventClientWrapper remoteEventClientWrapper = new RemoteEventClientWrapper() {
+
+      @Override
+      public void onParaInfosReceived(String noteId, String paragraphId,
+          Map<String, String> infos) {
+        if (infos != null) {
+          paraIdToInfosMap.put(paragraphId, infos);
+        }
+      }
+
+      @Override
+      public void onMetaInfosReceived(Map<String, String> infos) {
+      }
+    };
     context = new InterpreterContext("note", "id", null, "title", "text",
         new AuthenticationInfo(),
         new HashMap<String, Object>(),
         new GUI(),
         new AngularObjectRegistry(intpGroup.getId(), null),
         new LocalResourcePool("id"),
         new LinkedList<InterpreterContextRunner>(),
-        new InterpreterOutput(null));
+        new InterpreterOutput(null)) {
+        
+        @Override
+        public RemoteEventClientWrapper getClient() {
```

Comment:
@Leemoonsoo Regarding the refactoring, I not only mean `RemoveEventClient`, but also other interpreter related components. Like `InterpreterFactory`, `InterpreterSetting`, `InterpreterGroup`, `RemoteInterpreter` and etc. I just feel the boundary and responsibility of these classes/components is not clear for now. `InterpreterFactory` is too complicated and should delegate many work to `InterpreterSetting`. And currently there's no single component to manage the lifecycle of interpreters. I know refactoring is a little risky and may affect functions. So I think maybe we put this as part of 0.8.  If you think this is feasible and necessary. I can write a more detailed document and gather more feedback from communities.  What do you think ?